paddle/fluid/operators/setitem_op.h

```diff
@@ -0,0 +1,199 @@
+      value_name = "fp32_values";
+      break;
+    case framework::proto::VarType::INT64:
+      value_name = "int64_values";
+      break;
+    default:
+      PADDLE_THROW(platform::errors::Unimplemented(
+          "Unsupported data type(code %d) for Setitem operator, only "
+          "supports bool, int32, float32 and int64.",
+          data_type));
+  }
+  return value_name;
+}
+
+inline framework::DDim GetSliceDims(const framework::DDim in_dims,
+                                    const std::vector<int> axes,
+                                    const std::vector<int> starts,
+                                    const std::vector<int> ends) {
+  framework::DDim slice_dims(in_dims);
+
+  for (size_t i = 0; i < axes.size(); ++i) {
+    int axis = axes[i];
+    int dim_value = in_dims[axis];
+
+    int start = starts[i] < 0 ? (starts[i] + dim_value) : starts[i];
+    int end = ends[i] < 0 ? (ends[i] + dim_value) : ends[i];
+    start = std::max(start, 0);
+    end = std::min(end, dim_value);
+
+    PADDLE_ENFORCE_GT(end, start, platform::errors::InvalidArgument(
+                                      "end should greater than start, but "
+                                      "received end = %d, start = %d",
+                                      end, start));
+    slice_dims[axis] = end - start;
+  }
+  return slice_dims;
+}
+
+template <typename DeviceContext, typename T>
+class SetitemKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const {
+    const int rank = ctx.Output<framework::LoDTensor>("Out")->dims().size();
+
+    switch (rank) {
+      case 1:
+        SetitemCompute<1>(ctx);
+        break;
+      case 2:
+        SetitemCompute<2>(ctx);
+        break;
+      case 3:
+        SetitemCompute<3>(ctx);
+        break;
+      case 4:
+        SetitemCompute<4>(ctx);
+        break;
+      case 5:
+        SetitemCompute<5>(ctx);
+        break;
+      case 6:
+        SetitemCompute<6>(ctx);
+        break;
+    }
+  }
+
+ private:
+  template <size_t D>
+  void SetitemCompute(const framework::ExecutionContext& ctx) const {
+    auto* in = ctx.Input<framework::LoDTensor>("Input");
+    auto* out = ctx.Output<framework::LoDTensor>("Out");
+
+    auto dtype =
+        static_cast<framework::proto::VarType::Type>(ctx.Attr<int>("dtype"));
+    auto axes = ctx.Attr<std::vector<int>>("axes");
+    auto starts = ctx.Attr<std::vector<int>>("starts");
+    auto ends = ctx.Attr<std::vector<int>>("ends");
+    auto shape = ctx.Attr<std::vector<int>>("shape");
+    auto* value_tensor = ctx.Input<framework::LoDTensor>("ValueTensor");
+
+    auto in_dims = in->dims();
+    auto value_dims = framework::make_ddim(shape);
+    auto slice_dims = GetSliceDims(in_dims, axes, starts, ends);
+
+    auto place = ctx.GetPlace();
+    auto& eigen_place =
+        *ctx.template device_context<DeviceContext>().eigen_device();
+
+    out->ShareDataWith(*in);
```

Comment:
We are not allowed to ShareDataWith output Tensor, reference: https://www.paddlepaddle.org.cn/documentation/docs/zh/advanced_guide/addon_development/new_op/op_notes.html#sharedatawith

What's the purpose here?

Reply:
Here, input and output is the same Paddle.Tensor, so `output->ShareDataWith(in)` is called.

If deleting `input`,  checking rank can not be at **Compile-Time**